OpenRA.Mods.Common/AI/HackyAI.cs

```diff
@@ -381,52 +381,16 @@ public void Activate(Player p)
 				resourceTypeIndices.Set(tileset.GetTerrainIndex(t.TerrainType), true);
 		}
 
-		// TODO: Possibly give this a more generic name when terrain type is unhardcoded
-		public bool EnoughWaterToBuildNaval()
+		public bool IsAreaAvailable<T>(int radius, HashSet<string> terrainTypes)
 		{
-			var baseProviders = World.ActorsHavingTrait<BaseProvider>()
-				.Where(a => a.Owner == Player);
-
-			foreach (var b in baseProviders)
-			{
-				// TODO: Properly check building foundation rather than 3x3 area
```

Comment:
You need to keep/carry over this comment - this is about the shortcomings of relying on Util.AdjacentCells which you're still using.

Reply:
ok...